walk.go

```
@@ -0,0 +1,75 @@
+package straw
+
+import (
+	"errors"
+	"os"
+	"path/filepath"
+)
+
+// SkipDir is used as a return value from WalkFuncs to indicate that
+// the directory named in the call is to be skipped. It is not returned
+// as an error by any function.
+var SkipDir = errors.New("skip this directory")
+
+// WalkFunc is the type of the function called for each file or directory
+// visited by Walk. The path argument contains the argument to Walk as a
+// prefix; that is, if Walk is called with "dir", which is a directory
+// containing the file "a", the walk function will be called with argument
+// "dir/a". The info argument is the os.FileInfo for the named path.
+//
+// If there was a problem walking to the file or directory named by path, the
+// incoming error will describe the problem and the function can decide how
+// to handle that error (and Walk will not descend into that directory). In the
+// case of an error, the info argument will be nil. If an error is returned,
+// processing stops. The sole exception is when the function returns the special
+// value SkipDir. If the function returns SkipDir when invoked on a directory,
+// Walk skips the directory's contents entirely. If the function returns SkipDir
+// when invoked on a non-directory file, Walk skips the remaining files in the
+// containing directory.
+type WalkFunc = func(string, os.FileInfo, error) error
+
+func walk(store StreamStore, path string, info os.FileInfo, walkFn WalkFunc) error {
+
+	if !info.IsDir() {
+		return walkFn(path, info, nil)
+	}
+
+	fileInfos, err := store.Readdir(path)
+	err1 := walkFn(path, info, err)
+
+	if err != nil || err1 != nil {
+		return err1
```

Won't this hide the error returned by `ReadDir`?

---

It's doing the same as in the stdlib.  However, during my copy/paste, I removed the comments. Here is the original. https://github.com/golang/go/blob/master/src/path/filepath/path.go#L366

---

Ah interesting